### ISSUE TITLE
Zenodo author list

### ISSUE THREAD

**Issue by lostanlen** [librosa/librosa#1659]
This is not an issue with the code base but with our Zenodo integration.

We've just released librosa v0.10.rc0 on PyPI, and that triggered a new Zenodo object:
https://zenodo.org/record/7618817

When reviewing it i noticed that the list of authors was surprisingly small, with some important contributors missing:
* dafaronbi 
* iranroman
* juliawilkins 
* faroit
* matangover
* Embot2000
* bshall
* beasteers
* ...

What seems to happen is that Zenodo applies a threshold on the number of commits ever made to the package. But some contributions were either squashed or coauthored with someone, so they don't look like they count as much.
https://github.com/librosa/librosa/graphs/contributors

Librosa has had 100+ contributors in its history (yay!), so i understand that it's impractical to list everyone all the time. But the current rule (6 commits or more, IIUC) looks unfair to me. Without the people i just listed, we wouldn't have multichannel processing, pYIN, perceptual loudness weighting, non-A440 tuning, advanced I/O documentation, etc.

So i'm opening this issue to discuss collectively about what would be a good policy for authorship credit in Zenodo. I am not talking about `AUTHORS.md`, which is fine. Some options i can see right now, just to kickstart the debate

(1) Replace number of commits by number of added LOCs (green), or added LOCs + removed LOCs (green + red). Either set a minimum number (e.g., 100) or restrict the author list to top-K (e.g., K=10).

(2) Include everyone who has authored at least one commit since the last stable release. Exclude everyone else, even if they have contributed a lot to previous releases.

(3) Let Brian decide on who's deserving of coauthorship credit, based on various factors: code but also issues, review, comments, etc.

Let me know what you think.

**Comment by bmcfee** [comment]
Thanks for raising this!  I'll admit that I had no idea that zenodo was silently dropping authors from the list behind the scenes.

I tried to dig around a bit, and couldn't find much information on how exactly zenodo determines the author list, but my hunch is that it's not a commit threshold but rather a "top-K authors" threshold according to github's automatically generated contributor list.

This is problematic on a lot of fronts, not the least of which is that it de facto rules out contributors that do not directly author commits (eg @lostanlen ).  Also problematic are authors of giant PR's that get squash-merged into a single commit, eg

@dafaronbi 
![image](https://user-images.githubusercontent.com/1190540/217625918-d58cb085-fb8e-4c85-bf9c-cc131dfc4b1b.png)

@matangover 
![image](https://user-images.githubusercontent.com/1190540/217626027-5e18e433-76bb-41f1-ae55-e6e2dede5027.png)

So I think this is definitely something we want to rectify before the 0.10.0 final release.

After a bit of digging, I found some technical guidelines on how to fix this: https://developers.zenodo.org/#other-questions-on-harvesting

TLDR: we need to make a `.zenodo.json` file in the repo and populate the authors field appropriately.

This now raises the question of what our policy *should* be for authorship in a release.  This can be an extremely thorny governance question, and I don't expect a "simple" answer.  There are plenty of reasons why a heuristic like lines-modified or commits/pr's merged will not work, and we'll need some manual intervention to maintain the list.

Now this makes things complicated, as there's a human in the loop, and we should have some transparency and accountability for how these decisions get made.  Before proposing any ideas of my own, I'd like to solicit suggestions from anyone/everyone, especially if there are good examples we can draw on from other packages!

**Comment by bmcfee** [comment]
I asked around on the scientific python discord, and the general sense I'm getting is that we should populate all contributors in .zenodo.json, and do a bit of manual curation for each release to keep in sync with AUTHORS.md (since I'm curating that anyway when updating the changelog).

I think this reasonably fixes the attribution issue, but there's still a question of priority.

I don't necessarily believe there's a single correct ordering of contributors by importance.  However, memory is linear, and we need to pick *some* ordering, and it should ideally correlate with *some* notion of importance.

@lostanlen had suggested ranking by red+green as a starting point, that is, the sum of lines added and deleted by a contributor across all merged PRs.  This seems plausible to me, but github doesn't easily expose that (we can get adds or deletes, but not both). As a proxy, deletes is actually pretty good, as they typically track adds anyway, and if an author manages to delete a bunch of things and still get merged, they've probably done something right. :grin: 

Example:
https://github.com/librosa/librosa/graphs/contributors?from=2012-10-14&to=2023-02-09&type=d

This is definitely not perfect, of course.   I can think of numerous contributors who made small, but important changes.  There's also still the issue of people whose contributions are not well captured by git logs... this will likely require a thumb on the scale at some point.

**Comment by bmcfee** [comment]
I took a first swipe at this, by which I mean, I had chatgpt write a script to compile this data for me :laughing: 

With a bit of tweaking, it came out reasonably.  See https://github.com/librosa/librosa/blob/35b735891d061b6f6ef51ec3e80762d980d91f28/scripts/create_author_list.py

I realized that it would be helpful to somehow include duration of contributions as well here.  The implementation above sorts contributors by `((lines added) + (lines deleted)) / (1 + max_week - min_week)`.  I'm not married to this ordering at all, and LIS earlier, we probably want to tweak this a little to factor in non-code contributions. 

So please take this as a "first draft" and not at all final ordering:

<details>

<summary> Automatically generated author index </summary>

```
bmcfee                        |	0.0006 lines added/deleted per week
mattmcvicar                   |	1e-05 lines added/deleted per week
dafaronbi                     |	1e-05 lines added/deleted per week
iranroman                     |	1e-05 lines added/deleted per week
matangover                    |	1e-05 lines added/deleted per week
stefan-balke                  |	6e-06 lines added/deleted per week
scjs                          |	4e-06 lines added/deleted per week
SuperKogito                   |	2e-06 lines added/deleted per week
craffel                       |	2e-06 lines added/deleted per week
lostanlen                     |	2e-06 lines added/deleted per week
bshall                        |	2e-06 lines added/deleted per week
danagilliann                  |	2e-06 lines added/deleted per week
cwitkowitz                    |	2e-06 lines added/deleted per week
fzalkow                       |	2e-06 lines added/deleted per week
urinieto                      |	1e-06 lines added/deleted per week
dpwe                          |	1e-06 lines added/deleted per week
jack-alexander-ie             |	1e-06 lines added/deleted per week
kyungyunlee                   |	1e-06 lines added/deleted per week
beasteers                     |	1e-06 lines added/deleted per week
Embot2000                     |	1e-06 lines added/deleted per week
carlthome                     |	1e-06 lines added/deleted per week
faroit                        |	9e-07 lines added/deleted per week
rabitt                        |	9e-07 lines added/deleted per week
ZiyaoWei                      |	8e-07 lines added/deleted per week
ajweiss                       |	8e-07 lines added/deleted per week
ebattenberg                   |	8e-07 lines added/deleted per week
keunwoochoi                   |	6e-07 lines added/deleted per week
r9y9                          |	5e-07 lines added/deleted per week
Cortexelus                    |	5e-07 lines added/deleted per week
AlexMetsai                    |	5e-07 lines added/deleted per week
levkk                         |	4e-07 lines added/deleted per week
stefansjs                     |	4e-07 lines added/deleted per week
juliawilkins                  |	4e-07 lines added/deleted per week
JuliaWilkinsSonos             |	4e-07 lines added/deleted per week
pfriesch                      |	4e-07 lines added/deleted per week
Asmitha-K                     |	4e-07 lines added/deleted per week
onkyo14taro                   |	3e-07 lines added/deleted per week
s-t-e-v-e-n-k                 |	3e-07 lines added/deleted per week
bluenote10                    |	3e-07 lines added/deleted per week
QEDan                         |	3e-07 lines added/deleted per week
achabotl                      |	2e-07 lines added/deleted per week
cghawthorne                   |	2e-07 lines added/deleted per week
cr2007                        |	2e-07 lines added/deleted per week
EmmaZhao0413                  |	2e-07 lines added/deleted per week
wendywang3315                 |	2e-07 lines added/deleted per week
dofuuz                        |	2e-07 lines added/deleted per week
juanis2112                    |	2e-07 lines added/deleted per week
futurulus                     |	2e-07 lines added/deleted per week
viktorandreevichmorozov       |	2e-07 lines added/deleted per week
kianelbo                      |	1e-07 lines added/deleted per week
nullmightybofo                |	1e-07 lines added/deleted per week
ajweiss-dd                    |	1e-07 lines added/deleted per week
P-bibs                        |	1e-07 lines added/deleted per week
BdeGraaff                     |	1e-07 lines added/deleted per week
cclauss                       |	1e-07 lines added/deleted per week
joshmoore                     |	1e-07 lines added/deleted per week
dorukhansergin                |	9e-08 lines added/deleted per week
romi1502                      |	8e-08 lines added/deleted per week
naktinis                      |	8e-08 lines added/deleted per week
beantowel                     |	7e-08 lines added/deleted per week
tae898                        |	6e-08 lines added/deleted per week
jpaasen                       |	6e-08 lines added/deleted per week
joonlim9                      |	5e-08 lines added/deleted per week
stefempath                    |	5e-08 lines added/deleted per week
Isn0gud                       |	4e-08 lines added/deleted per week
alexmalins                    |	4e-08 lines added/deleted per week
kant                          |	4e-08 lines added/deleted per week
NumesSanguis                  |	3e-08 lines added/deleted per week
LorenzNickel                  |	3e-08 lines added/deleted per week
i-aki-y                       |	3e-08 lines added/deleted per week
wkcn                          |	3e-08 lines added/deleted per week
nehz                          |	3e-08 lines added/deleted per week
timgates42                    |	2e-08 lines added/deleted per week
mvollrath                     |	2e-08 lines added/deleted per week
stuartarchibald               |	2e-08 lines added/deleted per week
woodshop                      |	2e-08 lines added/deleted per week
encukou                       |	2e-08 lines added/deleted per week
Xiao-Ming                     |	2e-08 lines added/deleted per week
WuYiming6526                  |	2e-08 lines added/deleted per week
alastair                      |	2e-08 lines added/deleted per week
polyrhythmatic                |	2e-08 lines added/deleted per week
voodoohop                     |	2e-08 lines added/deleted per week
mattiadg                      |	1e-08 lines added/deleted per week
helminzo                      |	1e-08 lines added/deleted per week
c-jg                          |	1e-08 lines added/deleted per week
Abduttayyeb                   |	1e-08 lines added/deleted per week
tcelox                        |	9e-09 lines added/deleted per week
toddrme2178                   |	9e-09 lines added/deleted per week
ZviBaratz                     |	6e-09 lines added/deleted per week
masguit42                     |	6e-09 lines added/deleted per week
zhuangzi926                   |	6e-09 lines added/deleted per week
cashlo                        |	6e-09 lines added/deleted per week
pcampr                        |	6e-09 lines added/deleted per week
esemeniuc                     |	6e-09 lines added/deleted per week
monsij                        |	6e-09 lines added/deleted per week
shayenne                      |	6e-09 lines added/deleted per week
piem                          |	6e-09 lines added/deleted per week
hojinlee                      |	6e-09 lines added/deleted per week
waldyrious                    |	6e-09 lines added/deleted per week
jfsantos                      |	6e-09 lines added/deleted per week
```

</details>

There are a couple of quirks here, eg @juliawilkins counting twice because of two accounts linked on a commit log, or a few coauthored PRs being double-counted.  But these are easy enough to deal with.

**Comment by dpwe** [comment]
This all looks good except

dpwe                          |	1e-06 lines added/deleted per week

Meaning I’ve added one line in a million weeks / 20,000 years?

Damn.


  DAn.

**Comment by bmcfee** [comment]
> Meaning I’ve added one line in a million weeks / 20,000 years?

Hah - I think this is probably in units of seconds... so, up to a constant factor? :shrug: 

Upon further reflection though, I suspect that normalizing by active time is not a good idea.

**Comment by lostanlen** [comment]
Nice script! 🤖 

> Upon further reflection though, I suspect that normalizing by active time is not a good idea.

+1. If authors knew the rule, it would incentivize them to write one massive PR and then never contribute again for maintenance ...

**Comment by matangover** [comment]
> I had chatgpt write a script to compile this data for me

It got the units wrong 😄 

> Upon further reflection though, I suspect that normalizing by active time is not a good idea

Another +1, if someone has been contributing for a longer period of time, they shouldn't be penalized for that :)

**Comment by bmcfee** [comment]
Updating here - there's a proposed .zenodo.json list of all authors I could identify over in #1666 .  My plan is to make a 0.10.0rc1 release just to test-drive the zenodo integration and make sure everything works, but now would be a great time for feedback.

**Comment by bmcfee** [comment]
I just pushed rc1 to test the zenodo integration, and it appears to have worked flawlessly:

https://zenodo.org/record/7650421

![image](https://user-images.githubusercontent.com/1190540/219698124-ce70b941-ec86-4351-870e-325260c0c954.png)

**Comment by lostanlen** [comment]
Great!
At the same time, i'm having second thoughts on the value of DOI's for release candidates. At best they are a way for us to test that DOI integration works, from a technical standpoint.
But from a bibliometry standpoint, i think i would be really nice to have a single DOI for the last release candidate and the stable release, since they're really the same "Digital Object". Not sure if that's doable nor how you all feel about it.

**Comment by bmcfee** [comment]
> At best they are a way for us to test that DOI integration works, from a technical standpoint.

Yes, that's exactly right.



> a single DOI for the last release candidate and the stable release, since they're really the same "Digital Object".

Zenodo doesn't do that exactly.  There's one global DOI for the project that serves as a redirect to the latest version, and each version gets its own DOI.  I'm reasonably happy with that approach since it's simple and does the right thing almost always.  (The exception being in our current time-frame between an RC and a final.)
